client/state/reader/streams/reducer.js

```diff
@@ -16,50 +17,104 @@ import {
 	READER_STREAMS_UPDATES_RECEIVE,
 	READER_STREAMS_SELECT_NEXT_ITEM,
 	READER_STREAMS_SELECT_PREV_ITEM,
+	READER_STREAMS_SHOW_UPDATES,
 } from 'state/action-types';
-import { keyToString, keyForPost, keysAreEqual } from 'reader/post-key';
+import { keysAreEqual } from 'reader/post-key';
 
 /*
  * Contains a list of post-keys representing the items of a stream.
  */
 export const items = ( state = [], action ) => {
 	switch ( action.type ) {
 		case READER_STREAMS_PAGE_RECEIVE:
-			const { posts } = action.payload;
-			const postKeys = posts.map( keyForPost );
+			const { streamItems, gap } = action.payload;
 
-			const newState = uniqBy( [ ...state, ...postKeys ], keyToString );
+			let nextState;
+			if ( !! gap ) {
+				const beforeGap = takeWhile( state, postKey => ! keysAreEqual( postKey, gap ) );
+				const afterGap = takeRightWhile( state, postKey => ! keysAreEqual( postKey, gap ) );
 
-			// only return newState if it has actually been modified
-			if ( newState.length > state.length ) {
-				return newState;
+				// create a new gap if we still need one
+				let nextGap = [];
+				const from = gap.from;
+				const to = moment( last( streamItems ).date );
+				if ( ! from.isSame( to ) ) {
+					nextGap = [ { isGap: true, from, to } ];
+				}
+
+				nextState = [ ...beforeGap, ...streamItems, ...nextGap, ...afterGap ];
+			} else {
+				nextState = [ ...state, ...streamItems ];
 			}
+
+			nextState = uniqWith( nextState, keysAreEqual );
+
+			if ( nextState.length === state.length ) {
```

Comment:
this doesn't seem safe...

Reply:
thought maybe it is. would it catch the case where a gap is replaced with one post?

Reply:
Thats a good point, I can make this optimization safer and always return `nextState`